packages/cli/src/util/dev/server.ts

```diff
@@ -1478,6 +1478,29 @@ export default class DevServer {
             'content-length',
             'transfer-encoding',
           ]);
+
+          const overriddenHeaders = middlewareRes.headers.get(
+            'x-middleware-override-headers'
+          );
+          if (overriddenHeaders) {
+            for (const name of overriddenHeaders.split(',')) {
+              if (skipMiddlewareHeaders.has(name)) {
+                continue;
+              }
+
+              const valueKey = `x-middleware-request-${name}`;
+              const newValue = middlewareRes.headers.get(valueKey);
+              if (newValue) {
+                req.headers[name] = newValue;
+              } else {
+                delete req.headers[name];
+              }
+
+              middlewareRes.headers.delete(valueKey);
+            }
+            middlewareRes.headers.delete('x-middleware-override-headers');
+          }
```

Comment:
do we want to extract this into a simple to understand function, that can easily be unit testable? `applyOverriddenHeaders(middlewareRes.headers)` <picture data-single-emoji=":eyeses:" title=":eyeses:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..JW-BeQw9Hre77NnW.7T6lj52pfWd_M8Zt3NfArpwU4hze2QfPY2PJjKcG8V6YV8qqc0E5WqHjIbcmDIX4uLIHx7A2247njl_f2IFfdZjZNHNXXzK8xF1uCxDHJnufir0-QkcGdQ.JKTKmpzdxmnWHjeZJb3-9w" alt=":eyeses:" align="absmiddle"></picture>

Reply:
I agree. Some unit tests for the new logic would be great in this or a follow-up PR.